src/microsim/schema/modality/_simple.py

```diff
@@ -39,22 +43,39 @@ def render(
         settings: Settings,
         xp: NumpyAPI | None = None,
     ) -> xrDataArray:
-        if truth.ndim > 4 or Axis.F not in truth.dims:
-            raise NotImplementedError(
-                "At this stage, we only support rendering 3D or 4D data with a "
-                "fluorophore dimension."
-            )
+        convolved = 0
+        for fluor_idx in range(truth.sizes[Axis.F]):
+            convolved_fluor = 0
+            for bin_idx in range(truth.sizes[Axis.W]):
+                binned_flux = truth.isel({Axis.W: bin_idx, Axis.F: fluor_idx})
+                if np.isnan(np.sum(binned_flux.data)):
+                    # NOTE: there can be bins for which there is no data in one of the
+                    #  fluorophores
+                    continue
+                em_wvl = binned_flux[Axis.W].values.item().mid * UnitRegistry().nm
+                psf = self.psf(
+                    truth.attrs["space"],
+                    channel,
+                    objective_lens,
+                    settings,
+                    xp,
+                    em_wvl=em_wvl,
+                )
+                convolved_fluor += xp.fftconvolve(
+                    binned_flux.isel({Axis.C: 0}), psf, mode="same"
+                )
+            convolved += convolved_fluor
 
-        xp = NumpyAPI.create(xp)
-        # convert label dimension to photon flux
-        psf = self.psf(truth.attrs["space"], channel, objective_lens, settings, xp)
-        convolved = [
-            xp.fftconvolve(truth.isel({Axis.F: f}).data, psf, mode="same")
-            for f in range(truth.sizes[Axis.F])
-        ]
-        img = xp.stack(convolved)
-
-        return DataArray(img, dims=truth.dims, coords=truth.coords, attrs=truth.attrs)
+        return DataArray(
```

Comment:
it would also be acceptable here to return a 4D array `(C,Z,Y,X)` where the C is just a singleton dimension.  the "benefit" there is that it retains the coordinate information of which channel resulted in this image.  
in fact, it would even be possible to retain the wavelength and fluorophore info in [non-dimensional coordinates](https://docs.xarray.dev/en/v2023.06.0/user-guide/data-structures.html#coordinates).  not necessary here (can do it later), but just a good reminder of how xarray lets you preserve metadata and provenance.

Reply:
Makes sense. Added.